server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1440,17 +1439,13 @@ int runTranslogRecovery(Engine engine, Translog.Snapshot snapshot, Engine.Operat
                     default:
                         throw new AssertionError("Unknown result type [" + result.getResultType() + "]");
                 }
-
-                opsRecovered++;
-                onOperationRecovered.run();
+            } catch (IOException e) {
+                throw e;
             } catch (Exception e) {
-                if (ExceptionsHelper.status(e) == RestStatus.BAD_REQUEST) {
```

Comment:
We will need to keep this for `origin == LOCAL_TRANSLOG_RECOVERY`, at least until we have a way to repair primaries that are stuck in this situation. Otherwise we run the risk that primaries are unrecoverable if the mapping is not up-to-date (for whichever reason).

Reply:
Also means that the peer recovery PR should probably use Operation.Origin.LOCAL_RESET as origin.